### PR TITLE
Avoid realizing locations array when finding the symbol found in a particular location.

### PR DESCRIPTION
**Pull request by CyrusNajmabadi**
Saves around 35MB in a typing/lightbulb scenario: 

image

### CODE REVIEW

File: src/Compilers/CSharp/Portable/Compilation/SyntaxTreeSemanticModel.cs [hunk 1]
```diff
@@ -1774,19 +1774,31 @@ private Symbol GetDeclaredMember(NamespaceOrTypeSymbol container, TextSpan decla
...
+                // If, however, the location name is empty, then this was a bogus symbol created in an error recovery
+                // scenario.  Remember this symbol for the end in case we aren't able to find anything better to return.
+                var firstLocation = symbol.TryGetFirstLocation();
+                if (firstLocation != null && !firstLocation.SourceSpan.IsEmpty)
                 {
-                    if (loc.IsInSource && loc.SourceTree == this.SyntaxTree && declarationSpan.Contains(loc.SourceSpan))
+                    if (symbol.IsDefinedInSourceTree(this.SyntaxTree, declarationSpan))
+                        return symbol;
+                }
+                else
```
Comment (sharwell) [inline src/Compilers/CSharp/Portable/Compilation/SyntaxTreeSemanticModel.cs:1791]: 💭 It seems like this `else` is invalid (would prevent certain cases from locating a symbol.

**Review by sharwell — DISMISSED** [review]
❓ Is this pull request attempting to resolve an independent observation of #49769? If so, my recommendation would be to review the specific observations from that issue and target it directly. In particular, the most significant impact of this is a case where the symbol being looked up _does_ have multiple parts, and therefore might not be applicable for a fast path for items with only a single location.

File: src/Compilers/CSharp/Portable/Compilation/SyntaxTreeSemanticModel.cs [hunk 2]
```diff
@@ -1774,20 +1774,13 @@ private Symbol GetDeclaredMember(NamespaceOrTypeSymbol container, TextSpan decla
                     }
                 }
 
-                foreach (var loc in symbol.Locations)
+                if (symbol.HasLocationContainedWithin(this.SyntaxTree, declarationSpan, out var wasZeroWidthMatch))
```
Comment (CyrusNajmabadi) [inline src/Compilers/CSharp/Portable/Compilation/SyntaxTreeSemanticModel.cs:1777]: note: This is distinctly different from IsDefinedInSourceTree which does an intersects check, while this helper does a *contains* check.  The intersects check means that nodes *larger than* the decl-span passed in are considered matches.

**Comment by CyrusNajmabadi** [comment]
> ❓ Is this pull request attempting to resolve an independent observation of https://github.com/dotnet/roslyn/issues/49769?

No, that linked issue is about a n^2 algorithm somewhere.  The issue here is not the number of calls (since .Locations is cached), but simply that allocating the full locations array is extremely expensive.  For example, in compiler layer, since practically every type is within Global.Microsoft.CodeAnalysis, we have to alloc location arrays on every edit that contain the locations of all those namespaces across all files (as well as all the subnamespaces of those, which are also in many files).  This ends up being a full percent of all allocs in the scenario.

File: src/Compilers/CSharp/Portable/Symbols/Symbol.cs [hunk 3]
```diff
@@ -346,6 +346,36 @@ internal virtual LexicalSortKey GetLexicalSortKey()
...
+
+            wasZeroWidthMatch = false;
+            return false;
+        }
+
+        protected static bool IsLocationContainedWithin(Location loc, SyntaxTree tree, TextSpan declarationSpan, out bool wasZeroWidthMatch)
+        {
+            if (loc.IsInSource && loc.SourceTree == tree && declarationSpan.Contains(loc.SourceSpan))
+            {
+                wasZeroWidthMatch = loc.SourceSpan.IsEmpty && loc.SourceSpan.End == declarationSpan.Start;
```
Comment (ToddGrun) [inline src/Compilers/CSharp/Portable/Symbols/Symbol.cs:371]: >loc.SourceSpan.End == declarationSpan.Start [](http://example.com/codeflow?start=62&length=43)

nit: isn't this check unnecessary?

File: src/Compilers/CSharp/Portable/Symbols/Source/SourceNamespaceSymbol.cs [hunk 4]
```diff
@@ -99,6 +99,19 @@ public override ImmutableArray<Location> Locations
             }
         }
```
Comment (ToddGrun) [inline src/Compilers/CSharp/Portable/Symbols/Source/SourceNamespaceSymbol.cs:100]: Is there no need for this to implement TryGetFirstSymbol, or was that done in a different PR?

File: src/Compilers/CSharp/Portable/Symbols/Symbol.cs [hunk 5]
```diff
@@ -346,6 +346,36 @@ internal virtual LexicalSortKey GetLexicalSortKey()
         public Location GetFirstLocation()
             => TryGetFirstLocation() ?? throw new InvalidOperationException("Symbol has no locations");
 
+        /// <summary>
+        /// Determines if there is a location (see <see cref="Locations"/>) for this symbol whose span is in <paramref
+        /// name="tree"/> and is contained within <paramref name="declarationSpan"/>.  Subclasses can override this to
+        /// be more efficient if desired (especially if avoiding allocations of the <see cref="Locations"/> array is
+        /// desired).
+        /// </summary>
+        public virtual bool HasLocationContainedWithin(SyntaxTree tree, TextSpan declarationSpan, out bool wasZeroWidthMatch)
```
Comment (ToddGrun) [inline src/Compilers/CSharp/Portable/Symbols/Symbol.cs:355]: >declarationSpan [](http://example.com/codeflow?start=81&length=15)

This isn't necessarily a declarationSpan, right?

**Review by CyrusNajmabadi — COMMENTED** [review]
Comment

**Comment by CyrusNajmabadi** [comment]
@333fred @RikkiGibson

File: src/Compilers/CSharp/Portable/Symbols/Symbol.cs (hunk 5 above)
Comment (CyrusNajmabadi) [reply src/Compilers/CSharp/Portable/Symbols/Symbol.cs:358]: i'm just using the original name the caller was using :)

**Comment by CyrusNajmabadi** [comment]
@AlekseyTs @dotnet/roslyn-compiler  ptal.

**Comment by CyrusNajmabadi** [comment]
@333fred @RikkiGibson @jaredpar ptal.

File: src/Compilers/CSharp/Portable/Symbols/Symbol.cs [hunk 6]
```diff
@@ -349,6 +349,36 @@ public Location GetFirstLocation()
...
+            {
+                if (IsLocationContainedWithin(loc, tree, declarationSpan, out wasZeroWidthMatch))
+                    return true;
+            }
+
+            wasZeroWidthMatch = false;
+            return false;
+        }
+
+        protected static bool IsLocationContainedWithin(Location loc, SyntaxTree tree, TextSpan declarationSpan, out bool wasZeroWidthMatch)
```
Comment (jaredpar) [inline src/Compilers/CSharp/Portable/Symbols/Symbol.cs:370]: Why is this API on `Symbol` and not `SyntaxTree`? It seems purely about the latter so confused why it's here.

Comment (CyrusNajmabadi) [reply src/Compilers/CSharp/Portable/Symbols/Symbol.cs:370]: primarily because it's just called from something symbolic and really seems to be about a symbolic concept (how we think about symbol containment and spans).  I could move to SyntaxTree if you'd prefer :)

Comment (RikkiGibson) [reply src/Compilers/CSharp/Portable/Symbols/Symbol.cs:370]: Deja vu. I swear I made the same comment about a method like this recently, because the relationship between `loc` and `tree` felt unclear to me.

**Review by RikkiGibson — COMMENTED** [review]
Concept LGTM, would just feel more comfortable with the new API being on SyntaxTree rather than Symbol.

File: src/Compilers/CSharp/Portable/Symbols/Symbol.cs (hunk 6 above)
Comment (RikkiGibson) [reply src/Compilers/CSharp/Portable/Symbols/Symbol.cs:370]: I am open to keeping this method here, if we feel the coupling to "here's how we decide if a symbol is *located here*" is too strong. Consider if there are more distinct names which can be used for the "where we are looking" versus the "possible thing we found" parameters.

**Review by RikkiGibson — APPROVED** [review]
I am open to keeping `IsLocationContainedWithin` where it is, if the coupling to how symbol lookup works feels too strong. Consider if there are parameter names which could better distinguish "where we are looking" from "thing that might be here".

**Review by jaredpar — APPROVED** [review]
